### PR TITLE
Accelerator model state dict

### LINKED ISSUE #7470: API change, expose model's state_dict to `accelerator.training_type_plugin`
## 🚀 Feature

Currently, in `CheckpointConnector.dump_checkpoint`, we have

```
model = self.trainer.lightning_module

checkpoint = {
    'epoch': current_epoch,
    'global_step': global_step,
    'pytorch-lightning_version': pytorch_lightning.__version__,
    'state_dict': model.state_dict(),
}

```

so model's state dict is extracted here. However, let `accelerator.training_type_plugin` control the logic might make more sense especially for sharded plugin, we might need to access the local (i.e. sharded) state instead of the whole states.

### Motivation

https://github.com/PyTorchLightning/pytorch-lightning/pull/6152#discussion_r629087663

we would like to make customized model state dict for specific training type plugin, we could override the `training_type_plugin.on_save` method to modify the state dict, but this would cause duplicate call for extracting model state dict.

### Pitch

define a new method for `TrainingTypePlugin`

```
def state_dict(self) -> dict:
     model = self.lightning_module
     return model.state_dict()
```

and in `CheckpointConnector.dump_checkpoint`, 
```
checkpoint = {
    'epoch': current_epoch,
    'global_step': global_step,
    'pytorch-lightning_version': pytorch_lightning.__version__,
    'state_dict': self.trainer.accelerator.training_type_plugin.state_dict(),
}

```

### PR DESCRIPTION
**Pull request by shuyingsunshine21**
## What does this PR do?

Fixes #7470

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]
cc @awaelchli @tchaton

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -420,6 +420,12 @@ def optimizer_state(self, optimizer: Optimizer) -> Dict[str, Tensor]:
         """
         return getattr(self.training_type_plugin, 'optimizer_state', lambda x: x.state_dict())(optimizer)
 
+    def state_dict(self) -> Dict[str, Union[Any, Tensor]]:
```
[the remark this answers is not in the document]
Comment (shuyingsunshine21) [reply pytorch_lightning/accelerators/accelerator.py:423]: Oh, I just follow the same type defined below for `on_save`, but I think you are right, `Dict[str, Any]` should be enough

**Comment by shuyingsunshine21** [comment]
> do you see a good way to unit test the state dict being exposed from the accelerator?

will check if we already have some existing test for state dict, if not, i will come up with something.

**Review by awaelchli — APPROVED** [review]
